### DISCUSSION TITLE
updating state of AsyncNotifier triggers a build

### DISCUSSION THREAD

**Discussion by roubachof** [rrousselGit/riverpod#1876] [Q&A] [answered]
There is something I'm struggling to understand with the async notifier.
I have the following provider:

```dart
@riverpod
class ZoneStreetLamps extends _$ZoneStreetLamps {
  static final _log = LoggerFactory.logger('ZoneStreetLampsProvider');

  late StreetLampRemoteRepository _repository;

  @override
  Future<IList<StreetLamp>> build({required CityZone zone}) async {
    _log.i('build( $zone )');

    _repository = ref.read(streetLampRemoteRepositoryProvider);
    final lamps = await _repository.getList(zone);
    return lamps.sort(streetLampComparator);
  }

  Future addOrUpdate(StreetLamp streetLamp) async {
    _log.i('addOrUpdate( $streetLamp )');

    final updatedLamp = await _repository.addOrUpdate(streetLamp);

    await update((currentList) {
      final updatedList =
          currentList.updateById([updatedLamp], (item) => item.id);
      return currentList.length != updatedList.length
          ? updatedList.sort(streetLampComparator)
          : updatedList;
    });
  }

  Future remove(StreetLamp streetLamp) async {
    _log.i('remove( $streetLamp )');
    await _repository.remove(streetLamp);

    await update((currentList) {
      return currentList.removeWhere((element) => element.id == streetLamp.id);
    });
  }
}
```

I then have another provider that depends on this provider by watching it:

```dart
@riverpod
class LampList extends _$LampList {
  static final _log = LoggerFactory.logger('LampListProvider');

  late CityZone selectedZone;

  @override
  Future<IList<StreetLamp>> build() async {
    _log.i('build()');

    if (state.isRefreshing) {
      ref.invalidate(zoneStreetLampsProvider(zone: selectedZone));
    }

    selectedZone = await ref.watch(selectedZoneProvider.future);
    final list =
        await ref.watch(zoneStreetLampsProvider(zone: selectedZone).future);

    _log.listCount(list);
    return list;
  }

  Future addOrUpdate(StreetLamp streetLamp) async {
    _log.i('addOrUpdate( $streetLamp )');
    return ref
        .read(zoneStreetLampsProvider(zone: selectedZone).notifier)
        .addOrUpdate(streetLamp);
  }

  Future remove(StreetLamp streetLamp) async {
    _log.i('remove( $streetLamp )');
    return ref
        .read(zoneStreetLampsProvider(zone: selectedZone).notifier)
        .remove(streetLamp);
  }
}
```

Then when I call `() => lampListNotifier.remove(lamp)` when taping on a button I have the following log:

```
flutter: LampListProvider [I]  remove( Instance of '_$_StreetLamp' )
flutter: ZoneStreetLampsProvider [I]  remove( Instance of '_$_StreetLamp' )
flutter: StreetLampRemoteRepositoryMock [I]  remove( Instance of '_$_StreetLamp' )
flutter: LampListProvider [I]  build()
flutter: ZoneStreetLampsProvider [I]  build( CityZone(id: 1, name: Paris) )
flutter: StreetLampRemoteRepositoryMock [I]  getList( CityZone(id: 1, name: Paris) )
flutter: LampListProvider [I]  1 items returned
```

LampListProvider is rebuilt... But I just set the state to a new value through an update, so I don't understand why he is rebuilding the whole thing making therefore a useless call to the remote repository...

**Comment by rrousselGit** [comment]
You're modifying a provider that `LampList` depends on using `ref.watch`
So that `ref.watch` causes `build` to rerun, that's expected.

**Reply by roubachof** [reply]
But since I modify directly the state of the provider, why does it need to rebuild?
It defeats the purpose of updating the state of the provider if it rebuilds itself for nothing no?
Again I am missing something :)

I will try to explain what I want to achieve:

`ZoneStreetLamp` get first its value from a remote (http) service, so the State of the provider is initialized.
Now I want to add an item to my list of lamps.
I do it by updating the remote list, so my backend is in sync, with an update method:

```dart
final updatedLamp = await _repository.addOrUpdate(streetLamp);
```

And then I want to update my local cache by using the update method of the asyncNotifier.

```dart
await update((currentList) {
      final updatedList =
          currentList.updateById([updatedLamp], (item) => item.id);
      return currentList.length != updatedList.length
          ? updatedList.sort(streetLampComparator)
          : updatedList;
    });
```

Now my state is in sync with my remote http service, and I surely don't want it to rebuild since it will include an unnecessary http call.
In my mind I thought that updating the state will update the value, thus the providers depending on it will just get the new state.
Then I thought since the state changed, my provider watching the `ZoneStreetLamp` state will just get it straight from the state (the cached value)...

Is there a way to achieve this classic update remote, update cache pattern ?

**Comment by roubachof** [comment, marked as answer]
Ok, I think I understand now what happened :)
Setting the state manually in fact doesn't force a rebuild (it would be defeating the purpose of it :).
In fact I have a layer of "domain" providers, that are acting like a reactive cache for my whole app state.
If you have such kind of providers, you don't want them to be auto disposed.
 
So you need to add the `keepAlive` property in the `@Riverpod` attribute to be sure they won't be disposed when some widget rebuild will be triggered. 

```dart
@Riverpod(keepAlive: true)
class ZoneStreetLamps extends _$ZoneStreetLamps {
```
